### ISSUE TITLE
jacobian of boxminus for SE(3)/SO(3)

### ISSUE THREAD

**Issue by v01d** [strasdat/Sophus#179]
Hi,
I'm trying to compute the jacobian of log(inv(T1).T2), where T1,T2 are in SE(3) using Sophus. I would like to have this since I'm using ceres with Sophus. While there is a boxplus with its jacobian provided in this repo, jacobian of boxminus is not provided and it is non-trivial. I would apply chain rule and apply jacobian of log() and jacobian of composition and inversion, but could not find a simple expression for this. I found that in [OKVIS](https://github.com/ethz-asl/okvis), [VINS-Mono](https://github.com/HKUST-Aerial-Robotics/VINS-Mono) and [slamtb](https://github.com/joansola/slamtb), there appears to be the same math at to some point, but all implementations differ and are not commented. I've also looked into [these notes](https://github.com/jlblancoc/tutorial-se3-manifold), but neither the Jacobian of log(x) is provided, nor are the jacobians of pose composition and inversion expressed w.r.t. the quaternion.

In any case, I see that Sophus uses symbolically computed expressions for jacobians. Maybe it would not be that hard to do so for boxminus?

**Comment by jlblancoc** [comment]
Hi Matias,

Not a full answer but just some hints that may be useful:

> I've also looked into these notes, but neither the Jacobian of log(x) is provided

It's actually there, perhaps you accessed an archived older version?

• 9/May/2013: Added the Jacobian of the SO(3) logarithm map, in §10.3.2.

> nor are the jacobians of pose composition and inversion **expressed w.r.t. the quaternion.**

That's true, and I would like to address that in a future version of the technical report!

However, that, and what you want to do, could be better addressed IMO by the Jacobian chain rule, instead of trying to obtain a full direct symbolic expression. Sometimes the matrix multiplication (numeric) is a simpler and even faster way of computing them, put aside that it's less error prone...

In other words: in theory all you need is the report, but a few Jacobian matrices need to be multiplied. If I have spare time to do it next week I may update the report with the result, but pull requests if someone want to do it before would be welcome!

**Comment by v01d** [comment]
Hi,
thanks for the fast response. If I understand correctly, you suggest to include the quaternion-to-matrix jacobian w.r.t. the quaternion in the chain, so that I can simply re-use the log, composition and inversion jacobians expressed w.r.t the matrix?

**Comment by jlblancoc** [comment]
Exactly!

**Comment by jlblancoc** [comment]
On a second thought, that's what needs to be done if your poses are as quaternions. If you take derivatives wrt algebra elements se(3), those are 3-vectors, so the quaternion to matrix formula may be needed to find out the rotation matrices as needed by Jacobians, but I thing the Jacobian of the quat to matrix wouldn't be needed...

**Comment by v01d** [comment]
@strasdat Would it be possible to provide derivatives of log() for SO(3) and SE(3) with the python symbolic module as you do for exp()?

**Comment by dangut** [comment]
I don't know if it applies to your case, but it might be worth trying to refactor your code so that instead of putting translation and rotation together in an SE(3) object, they are treated separately as independent SO(3) and R3 elements (personally, I still haven't faced a problem where that can't be done). That has two advantages. Firstly, no coupling with rotation is introduced in the translation part of the exponential and log map, and thus translation boxminus and boxplus are trivially linear operators. Secondly, it seems that the Jacobian of the log map involves computing the Campbell-Baker-Hausdorff Formula which apparently does not have a closed form solution for SE(3); but it does for SO(3), related with what Forster et al. call the right Jacobian of SO(3) here https://arxiv.org/pdf/1512.02363.pdf (check Eq(9)).

If you opt for this option, I recommend also you to check this paper https://arxiv.org/pdf/1606.05285.pdf. But be aware that they use the convention of differentiating in the left tangent space. If you differentiate on the right tangent space the Jacobians are slightly different.

**Comment by v01d** [comment]
Thank you for your input! I'm actually currently using the approach you mention. Although, I wonder how much is lost with such approximation. 
I was unaware of the fact that log of SE(3) has no closed form. I will check back into the literature.

**Comment by jlblancoc** [comment]
As Forster et al. say in their paper above, the operation is actually a "retraction", one of its possible implementations being directly the matrix exp/log since they are Lie groups, but splitting the translation and rotation seems also valid. I also wonder in someone has done a systematic evaluation of whether the choice of retraction has any practical impact on accuracy...

Thanks for the great discussion!

**Comment by dangut** [comment]
You are welcome! Actually, there is this paper https://arxiv.org/pdf/1805.02543.pdf where Hovren et al. make a comparison of SE(3) splines vs split SO(3) and R3 splines for SfM. They show in the experiments that even in cases which would favor SE(3) (forward locked and sideways locked motions), split representation performs better. I think the intuition behind the results is that although both representations make different assumptions on the kind of motion, what will have the last word and more weight on how the optimized variables are coupled are the problem cost functions, and in that sense it might be better to have a more linear system.

**Comment by NikolausDemmel** [comment]
> I also wonder in someone has done a systematic evaluation of whether the choice of retraction has any practical impact on accuracy...

Yes, that'd be interesting indeed. Accuracy I'd expect to be the same, since all the Jacobians are the same independent of your retraction (apparently you should choose it such that it has the same Jacobian as the exponential at 0, namely identity) and close to the minimum the updates tend to 0 anyway and if your initialization is close enough to a minimum, you should converge to the same minimum regardless via which path exactly (but my intuition might be wrong here). There could be a difference in convergence speed, which would need to be traded off against the difference efficiency of computing the retraction.

> You are welcome! Actually, there is this paper https://arxiv.org/pdf/1805.02543.pdf where Hovren et al. make a comparison of SE(3) splines vs split SO(3) and R3 splines for SfM. They show in the experiments that even in cases which would favor SE(3) (forward locked and sideways locked motions), split representation performs better. I think the intuition behind the results is that although both representations make different assumptions on the kind of motion, what will have the last word and more weight on how the optimized variables are coupled are the problem cost functions, and in that sense it might be better to have a more linear system.

I just glanced at it, but it seems to me that this is more about interpolating trajectories rather than retractions to update manifold variables during optimization. I'd be surprised if you can conclude from that paper that the simplified retraction that decouples the update of translation from "delta \omega" is favourable over the full SE exp retraction in terms of accuracy (but I'm happy to be convinced otherwise).

**Comment by NikolausDemmel** [comment]
> I don't know if it applies to your case, but it might be worth trying to refactor your code so that instead of putting translation and rotation together in an SE(3) object, they are treated separately as independent SO(3) and R3 elements.

Just to clarify what I'm talking about in my last comment. When I talk about a retraction that decouples the update of the translation from the update of the rotation (as advocated e.g. in the mentioned Forster paper), I mean that you still represent your pose variables as proper SE(3) objects, with translation and orientation correlated. The only change is how you update your manifold variables in every step of Gauss Newton. If you look carefully at the retraction in the Forster paper (Eqn. 21), you notice that the translational update still depends on the current orientation R, just not on on the rotational update "delta phi", which it would for the full SE(3) exp retraction.

**Comment by dangut** [comment]
Possibly I am wrong but it doesn't seem to me that premultiplying by R in the retraction may have any effect. I didn't do the math but from my intuition I would say that the effect would be that every Jacobian involving a translation would be postmultiplied by its corresponding rotation which would mean that dp_{Forster} = R^T*dp, and then, the result is the same.

**Comment by v01d** [comment]
Regarding the closed form for left/right jacobian of SE(3), in "State Estimation for Robotics" from Barfoot, there appears to be a closed expression (pages 234-237). However, I believe this is the jacobian of the exponential, not the logarithm, AFAIK. 

To clarify, I need the jacobian of log(T_a^{-1}  (T_b exp(eps))), w.r.t. eps \in R^6 (in my case I actually have quaternion + translation, but I think the jacobian is easier this way). Applying chain, rule, I would only need jacobian of logarithm w.r.t a SE(3) transform, since jacobian of pose composition and SE(3) plus operator are known.

**Comment by NikolausDemmel** [comment]
But I think that one you can do directly with the right Jacobian of SE(3). You can check [Sola's report](http://www.iri.upc.edu/people/jsola/JoanSola/objectes/notes/kinematics.pdf) (186) and (187). That is on SO(3), but I believe the same principles of log, exp, right and left Jacobians apply equally to SE(3).

First, we call use the shorthand T = T_a^{-1} T_b, since that doesn't make a difference for the Jacobian in question.

d / d eps (log(T exp(eps)))) = d / d eps (log(T) + J_r^-1(log(T)) eps) = J_r^-1(log(T))

Or am I missing something?

Acutally, comming back to your original question.

> I'm trying to compute the jacobian of log(inv(T1).T2), where T1,T2 are in SE(3) using Sophus. I would like to have this since I'm using ceres with Sophus.

I'm curious why you would need this derivative for ceres? Isn't ceres' autodiff doing that for you?

**Comment by v01d** [comment]
Your approach sounds reasonable. I will look into it further and see if I can implement it.

Regarding your question, the point is that I want to implement analytical jacobians for performance reasons. Thus, I need to obtain these expressions myself.

**Comment by NikolausDemmel** [comment]
> Your approach sounds reasonable. I will look into it further and see if I can implement it.

Btw for `d/d eps (log(T_a exp(eps))^-1 T_b)` with the same approach I get `- J_l^-1 (log(T))` with the same `T`.

> Regarding your question, the point is that I want to implement analytical jacobians for performance reasons. Thus, I need to obtain these expressions myself.

Makes sense. You might wanna look at the code generation stuff using sympy in this library. With all the helpers from Hauke it should be quite simple for you to generate code for any Jacobians you need. And you can use ceres to numerically verify the Jacobians.

Check #183 for some initial work to generate Jacobians for log for SE(3) and SO(3).

**Comment by v01d** [comment]
I looked over your derivation and in facts seems correct. Indeed the same applies for SE(3) (you can look in eq 7.100a in Barfoot's book). So, this means we don't actually need jacobian of log() right? It actually only requires the closed form of the inverse of right jacobian of SE(3), which we can also get from Barfoot.

**Comment by NikolausDemmel** [comment]
Yes, that seem right. `J_r^-1` and maybe also `J_l^-1` for the derivate w.r.t. `T_a`.

**Comment by NikolausDemmel** [comment]
> Possibly I am wrong but it doesn't seem to me that premultiplying by R in the retraction may have any effect. I didn't do the math but from my intuition I would say that the effect would be that every Jacobian involving a translation would be postmultiplied by its corresponding rotation which would mean that dp_{Forster} = R^T*dp, and then, the result is the same.

I guess in practice it should not make much of a difference as far as I can see. If you drop the R in the update, you also drop it from the Jacobian and you can imagine to simply use a rotated coordinate system in the translational part of the manifold tangent space. Since det(R) is 1, it should also not affect how your optimization is conditioned.

Maybe the only point is that if you use a proper "retraction", the Jacobians are actually the same as if you would use the full SE(3) exponential in your pose update.

**Comment by versatran01** [comment]
@NikolausDemmel For a related question, if I were to use the retraction in Foster's paper
`(R Exp(d_phi), p + R d_p)`, then I would have to change line 23 in `local_parameterization_se3.hpp` 
to 
```
T_plus_delta.so3() = SO3d::exp(delta.head<3>());
T_plus_delta.translation() = T.translation() + T.so3() * delta.tail<3>();
```
and line 36 to
```
 jacobian.topLeftCorner<4,3>() = T.so3().Dx_this_mul_exp_x_at_0(), 
jacobian.bottomRightCorner<3, 3>() = T.so3();
```
is that correct?

**Comment by NikolausDemmel** [comment]
@versatran01, your first line is missing the `T.so3() *` part, i.e.

```
T_plus_delta.so3() = T.so3() * SO3d::exp(delta.tail<3>());
```

For the Jacobian, I believe that should actually be unchanged and equal to the Jacobian of the "full Se(3) update". AFAIK that is the defining property of a retraction.

But, you should also be able to compute it like you propose, and it should give you the exact same Jacobian (maybe you could verify this). However, you might have mixed up the order of your variables. I believe it should be topRightCorner for the rotational part and bottomLeftCorner for the translational part.

**Comment by versatran01** [comment]
Ah yes you are right, it’s a bit confusing that in delta it is upsilon theta and in T it is rotation translation. 

I will take a look at if they give the same exact jacobians.

**Comment by NikolausDemmel** [comment]
> I will take a look at if they give the same exact jacobians.

Great, please report back when you got some results.

**Comment by strasdat** [comment]
old...
